src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -140,9 +161,24 @@ private static float CosineSimilarityCore(ReadOnlySpan<float> x, ReadOnlySpan<fl
                     Vector128<float> xVec = Vector128.LoadUnsafe(ref xRef, (uint)i);
                     Vector128<float> yVec = Vector128.LoadUnsafe(ref yRef, (uint)i);
 
-                    dotProductVector += xVec * yVec;
-                    xSumOfSquaresVector += xVec * xVec;
-                    ySumOfSquaresVector += yVec * yVec;
+                    if (Fma.IsSupported)
+                    {
+                        dotProductVector = Fma.MultiplyAdd(xVec, yVec, dotProductVector);
+                        xSumOfSquaresVector = Fma.MultiplyAdd(xVec, xVec, xSumOfSquaresVector);
+                        ySumOfSquaresVector = Fma.MultiplyAdd(yVec, yVec, ySumOfSquaresVector);
+                    }
+                    else if (AdvSimd.IsSupported)
+                    {
+                        dotProductVector = AdvSimd.FusedMultiplyAdd(dotProductVector, xVec, yVec);
+                        xSumOfSquaresVector = AdvSimd.FusedMultiplyAdd(xSumOfSquaresVector, xVec, xVec);
+                        ySumOfSquaresVector = AdvSimd.FusedMultiplyAdd(ySumOfSquaresVector, yVec, yVec);
+                    }
+                    else
+                    {
+                        dotProductVector += xVec * yVec;
```

Comment:
do we care here about the fact these paths are not the same in terms of IEEE754?

Reply:
> My general impression that FMA doesn't improve things a lot to trade determinism for

What determinism are you referring to? Do you mean across machines?

I thought it was desirable in terms of gaining a more accurate answer by not losing accuracy due to intermediate rounding.

If it's not beneficial, we can close this, though then I wonder why we've gone through the trouble of exposing all of these FMA primitives.

Reply:
> What determinism are you referring to? Do you mean across machines?

Yep, however, if we can assume FMA is mostly always available we can just change the fallback to match its behavior - it will be slower but at least the result will be repeatable accross different platforms.

> If it's not beneficial

It's probably just my impression, but I can be wrong. Also, like you mentioned, it also provides better accuracy

Reply:
Right, the actual performance benefit of `FMA` is marginal outside of specific scenarios but it does change the maximum error of the 2 step result from `1 ULP` to `0.5 ULP` (Units in Last Place). This may or may not be a problem depending on whether its part of a converged result.

The main places it helps with performance is very heavy `multiply-add` chains where it can allow better utilization of the multiplier unit. It can cause a perf regression or no improvement in some cases like `(a * b) + (c * d)” where all data is simultaneously ready due to the data dependence it introduces on the FMA